alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,173 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Error;
+use std::os::raw::c_void;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+use mio_extras::channel::{channel, Receiver};
+
+use winapi::shared::ntdef::{BOOLEAN, HANDLE, PVOID};
+use winapi::um::winbase::{RegisterWaitForSingleObject, UnregisterWait, INFINITE};
+use winapi::um::winnt::{WT_EXECUTEINWAITTHREAD, WT_EXECUTEONLYONCE};
+
+use crate::tty::ChildEvent;
+
+/// WinAPI callback for `HandleWaitSignal`, unpacks Rust closure reference and calls it
+extern "system" fn child_exit_callback<F>(ctx: PVOID, timed_out: BOOLEAN)
+where
+    F: FnOnce() + Send,
+{
+    if timed_out != 0 {
+        return;
+    }
+
+    let callback: Box<F> = unsafe { Box::from_raw(ctx as *mut F) };
+    callback();
+}
+
+/// Represents closure attached to Win32 handle wait signal
+///
+/// This allows to fire a Rust callback when subprocess exits
+pub(crate) struct HandleWaitSignal {
+    wait_handle: AtomicPtr<c_void>,
+}
+
+impl HandleWaitSignal {
+    /// Registers an asynchronous closure to call when process under `child_handle` exits.
+    ///
+    /// The `on_exit` is called on Win32 threadpool thread so it should avoid
+    /// blocking calls. See [`WT_EXECUTEINWAITTHREAD` flag docs] for details.
+    ///
+    /// [`WT_EXECUTEINWAITTHREAD` flag docs]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerwaitforsingleobject
+    fn new<F>(child_handle: HANDLE, on_exit: F) -> Result<HandleWaitSignal, Error>
+    where
+        F: FnOnce() + Send,
+    {
+        let mut wait_handle: HANDLE = 0 as HANDLE;
+        let wait_notify = Box::new(on_exit);
+
+        let success = unsafe {
+            RegisterWaitForSingleObject(
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback::<F>),
+                Box::into_raw(wait_notify) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if 0 == success {
+            Err(Error::last_os_error())
+        } else {
+            Ok(HandleWaitSignal { wait_handle: AtomicPtr::from(wait_handle) })
+        }
+    }
+}
+
+impl Drop for HandleWaitSignal {
+    fn drop(&mut self) {
+        unsafe {
+            // Cancel WinAPI wait for child process exiting
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
```

Comment:
It has to be `Send` since it's being a part of `Term` struct moved between threads by the main event loop.

Otherwise, you get:
```
error[E0277]: `*mut std::ffi::c_void` cannot be sent between threads safely
   --> alacritty\src\main.rs:181:22
    |
181 |     let event_loop = EventLoop::new(Arc::clone(&terminal), event_proxy.clone(), pty, &config);
    |                      ^^^^^^^^^^^^^^ `*mut std::ffi::c_void` cannot be sent between threads safely
    |
    = help: within `alacritty_terminal::tty::windows::Pty<'_>`, the trait `std::marker::Send` is not implemented for `*mut std::ffi::c_void`   
    = note: required because it appears within the type `alacritty_terminal::tty::windows::child::HandleWaitSignal`
    = note: required because it appears within the type `alacritty_terminal::tty::windows::child::ChildProcessWatcher`
    = note: required because it appears within the type `alacritty_terminal::tty::windows::Pty<'_>`
    = note: required by `alacritty_terminal::event_loop::EventLoop::<T, U>::new`
```

Reply:
It has to be `Send` since it becomes part of `Term` that is shared between threads by the main event loop.

Otherwise you get:
```
error[E0277]: `*mut std::ffi::c_void` cannot be sent between threads safely
   --> alacritty\src\main.rs:181:22
    |
181 |     let event_loop = EventLoop::new(Arc::clone(&terminal), event_proxy.clone(), pty, &config);
    |                      ^^^^^^^^^^^^^^ `*mut std::ffi::c_void` cannot be sent between threads safely
    |
    = help: within `alacritty_terminal::tty::windows::Pty<'_>`, the trait `std::marker::Send` is not implemented for `*mut std::ffi::c_void`   
    = note: required because it appears within the type `alacritty_terminal::tty::windows::child::HandleWaitSignal`
    = note: required because it appears within the type `alacritty_terminal::tty::windows::child::ChildProcessWatcher`
    = note: required because it appears within the type `alacritty_terminal::tty::windows::Pty<'_>`
    = note: required by `alacritty_terminal::event_loop::EventLoop::<T, U>::new`
```

Reply:
Makes sense, thanks for the clarification.